runtime/onert/backend/cpu/ops/BinaryArithmeticLayer.cc

```diff
@@ -30,31 +30,54 @@ namespace ops
 namespace
 {
 
-template <nnfw::cker::BinaryArithmeticOpType arithmetic_type, typename T>
-void eval(const IPortableTensor *lhs, const IPortableTensor *rhs, IPortableTensor *output,
-          nnfw::cker::BinaryArithmeticOpParam op_params)
+template <nnfw::cker::BinaryArithmeticOpType arithmetic_type, typename T> struct Eval
 {
-  const auto lhs_shape = getTensorShape(lhs);
-  const auto rhs_shape = getTensorShape(rhs);
-  const bool need_broadcast = nnfw::cker::ProcessBroadcastShapes(lhs_shape, rhs_shape, &op_params);
-  if (need_broadcast)
+  nnfw::cker::Shape _lhs_shape;
+  nnfw::cker::Shape _rhs_shape;
+  nnfw::cker::Shape _output_shape;
+  nnfw::cker::BinaryArithmeticOpParam _op_params;
+  bool _need_broadcast;
+
+  Eval(const IPortableTensor *lhs, const IPortableTensor *rhs, IPortableTensor *output,
+       nnfw::cker::BinaryArithmeticOpParam op_params)
+      : _op_params(std::move(op_params))
   {
-    nnfw::cker::BroadcastBinaryArithmeticOp<arithmetic_type>(
-        op_params, lhs_shape, reinterpret_cast<const T *>(lhs->buffer()), rhs_shape,
-        reinterpret_cast<const T *>(rhs->buffer()), getTensorShape(output),
-        reinterpret_cast<T *>(output->buffer()));
-    return;
+    if (!output->is_dynamic())
+      updateCache(lhs, rhs, output);
   }
 
-  nnfw::cker::BinaryArithmeticOp<arithmetic_type>(
-      op_params, lhs_shape, reinterpret_cast<const T *>(lhs->buffer()), rhs_shape,
-      reinterpret_cast<const T *>(rhs->buffer()), getTensorShape(output),
-      reinterpret_cast<T *>(output->buffer()));
-}
+  void updateCache(const IPortableTensor *lhs, const IPortableTensor *rhs, IPortableTensor *output)
+  {
+    _lhs_shape.ReplaceWith(getTensorShape(lhs));
+    _rhs_shape.ReplaceWith(getTensorShape(rhs));
+    _output_shape.ReplaceWith(getTensorShape(output));
+    _need_broadcast = nnfw::cker::ProcessBroadcastShapes(_lhs_shape, _rhs_shape, &_op_params);
+  }
+
+  void operator()(const IPortableTensor *lhs, const IPortableTensor *rhs, IPortableTensor *output)
+  {
+    if (output->is_dynamic())
```

Comment:
@hyunsik-yoon Is there any case where tensor is dynamic during `compile` and becomes static during `inference`? If that case occurs, `_lhs_shape` will be used without initialize.

Reply:
According to https://github.com/Samsung/ONE/issues/4544#issuecomment-707460292, it can happen.

Reply:
Correct. BTW we don't have such test case yet so I am writing the case.

Reply:
Designing cache strategy for various cases will take time.

For simple start, how about not using (or updating) the cache if `output->is_dynamic()`?
In other words, when `output->is_dynamic()`, the code works just like the previous way (by calling `getTensorShape`).

I am not sure how many tensors you want to cache. If you eventually consider to expand the scope of cache to all tensors in a model, I tried to calculate an imaginary huge model where there is, e.g., 1000 operations and works with dynamic tensors (meaning hit ratio of cache for static tensor shape is low and simple caching strategy waste memory),
the cache size would be [ 2.5 tensors / op * 1000  op * 28Bytes (size of nnfw::cker::Shape) = 68KB ]

I guess that we can eventually provide complex cache strategy or provide some flag to invalidate caching for the imaginary model I mentioned.

Reply:
for https://github.com/Samsung/ONE/pull/4611#discussion_r504403760

> Correct. BTW we don't have such test case yet so I am writing the case.

it will take some time. please refer to https://github.com/Samsung/ONE/issues/4625

Reply:
With offline discussion with @hyunsik-yoon, I've noticed that caching tensor shape for dynamic output may slow down inference when model has dynamic tensor. I agree with his suggestion that removes `updateCache` when output is dynamic.

Reply:
@hyunsik-yoon Can you be more specific with the scenario when dynamic tensor becomes static again? As I see from code, static shapes are inferenced in compile-time only, and the only other way of shape inference is the dynamic one which sets a new shape with applyShape method (at least for binary arithmetic op) which again always sets dynamic flag of an output.

Reply:
@periannath This is more theoretic, as `updateCache` does things that are done anyway, the only slowdown is the single call to `is_dynamic()` and probably `cker::Shape` copy, but copy can be eliminated by making additional `getTensorShape` signature that takes `Shape` reference and fills existing object instead of creating a new one.

Reply:
> @hyunsik-yoon Can you be more specific with the scenario when dynamic tensor becomes static again? As I see from code, static shapes are inferenced in compile-time only, and the only other way of shape inference is the dynamic one which sets a new shape with applyShape method (at least for binary arithmetic op) which again always sets dynamic flag of an output.

@krayzemli In current implementation, you may be right. However, there is no rule and much reason that we forbid dynamic shape to static shape.

Reply:
@hyunsik-yoon
I want to clarify memory requirements:
From my perspective ~100 KB overhead is not a big deal in case of neural networks, I think it is a fare trade-off if it can improve performance. Are you agree or I misunderstand something?

Reply:
@krayzemli

> Can you be more specific with the scenario when dynamic tensor becomes static again? 

It depends on the result of https://github.com/Samsung/ONE/issues/4625.

> This is only theoretic, as updateCache does things that are done anyway, the only slowdown is the single call to is_dynamic() and probably Shape::ReplaceWith(Shape&&) rvalue-copy, but this copy can be eliminated by making additional getTensorShape signature that takes Shape& reference and fills existing object instead of creating a new one.

Assuming the following model (similar with the test model I am working on):
(note that `#0` means 0th tensor)

```
#0 = input of shape [n] and n is the number of input from user
#1 = const of value [20]
WHILE #1 > 0
  #1 = SUB (#1 , 1)
  #0 = ADD (#0, #0)
#2 (output) = #0
```

application code is
```
nnfw_prepare();

while (input is `q`)
   n = read from user
   nnfw_set_input_tensor(#0, [n])
   nnfw_run()
```

`n` changes over time. In addition there is a WHILE loop. So caching tensorshape for `ADD` is overhead in this scenario.
`SUB` is also similar. Cache hit ratio is almost 0% in this case.

Reply:
@binarman for https://github.com/Samsung/ONE/pull/4611#discussion_r505117981
I agree now. :-)
Maybe we will find ways to reduce memory when we started to get strong request to reduce memory later. :-)

Reply:
@hyunsik-yoon 
> So caching tensorshape for `ADD` is overhead in this scenario.
> `SUB` is also similar. Cache hit ratio is almost 0% in this case.

The only overhead here is that cached parameters are stored in Eval object local memory (so they can be reused on the cache hit) rather than on the stack (where they can't be reused). There is some very small overhead of Shape::ReplaceWith, but it can be easily eliminated (see https://github.com/krayzemli/ONE/tree/binShapeCache2 ). Since we don't have real dynamic networks for experimenting with, we can't actually measure performance impact of cache miss.

Reply:
@krayzemli 

My curiosity is not about its overhead but more about whether this line is meaningful or not.

IMHO, this line is will provide improvement when _previous_ `nnfw_run()`'s shape is same with those in the _next_ `nnfw_run()` (case A) 
but this line is not meaningful when those shapes are different most of time (case B). 

Case B will be some language model where input size varies whenever `nnfw_run()` is called like [in the example I mentioned](https://github.com/Samsung/ONE/pull/4611/files#r505173665).

I wonder what would be the real example for case A.
If we don't have a specific example for case A, IMHO there is no point to leave this line (regardless of how much  overhead it adds or not.) considering [YAGNI](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it).

If we don't know case A at this moment, IMHO how about removing this line and adds some comment like following?
```
// TODO Adds lines like the following when a model using dynamic tensors needs shape caching
// if (output->is_dynamic())
//      updateCache(lhs, rhs, output);
```

Reply:
This line updates (invalidates) the cache every run (since a tensor became dynamic, it is never static again), without this line dynamic networks become broken. Only static models benefit from the cache reuse,.

Reply:
@krayzemli Oh. I see. Then let's leave as is. :-)

Reply:
> This line updates (invalidates) the cache every run (since a tensor became dynamic, it is never static again), without this line dynamic networks become broken. Only static models benefit from the cache reuse,.

@krayzemli It seems you still assume once it become dynamic, it cannot be back to static. We may make a rule internally after discussion. However, I've never heard explicitly about this kind of rule. Thus, I don't think it is always true.

Reply:
@glisting To make tensor static again, static shape inferer must be rerun, and right now, there is no such mechanism in the runtime. Introducing such a mechanism will trigger a lot of redesign anyway. I think I could start such a redesign soon after my probation.

Reply:
@krayzemli 
I'll try to reword what @glistening said:
- We do not have agreement about dynamic shapes yet and we don't have any checking mechanism in runtime that will ensure properties of tensors.
- Many people are working/worked/will work on runtime so we can not be sure code will stay as it is today.

So we can not rely on 
> now, there is no such mechanism in the runtime

> Introducing such a mechanism will trigger a lot of redesign anyway

I am not sure this is hard to introduce, there are a lot of logic in data-flow executor, for example. I believe some magic related to shapes could be done there.

Anyway, if you want to redesign something - feel free to start a conversation, so everyone can understand your ideas =)

Reply:
@binarman Would adding a formal assert solve the problem then?

Reply:
@binarman Thank you for rewording to clarifying my statements.

@krayzemli Yes, if you can add `assert` here, it will make `onert` easy to debug in not-allowed but happened case.